### PR TITLE
DOC: fix CL tutorial to give same output from saved file and example

### LINKED ISSUE #12391: Constrained Layout tutorial needs some cleanup....
_Originally posted by @ImportanceOfBeingErnest in https://github.com/matplotlib/matplotlib/pull/12377#issuecomment-426807732_

The code does not seem to be runnable and does not give a sensible result. I used this code

```
fig, axs = plt.subplots(2, 2, constrained_layout=True)
for ax in axs.flatten()[:-1]:
    ax.plot(np.arange(10))
axs[1, 1].plot(np.arange(10), label='This is a plot')
leg = axs[1, 1].legend(loc='center left', bbox_to_anchor=(0.8, 0.5))
leg.set_in_layout(False)
wanttoprint = True
if wanttoprint:
    leg.set_in_layout(True)
    fig.do_constrained_layout(False)
    fig.savefig('outname.png', bbox_inches='tight')
plt.show()
```

i.e. the code from the [matplotlib 3.0 documentation](https://matplotlib.org/tutorials/intermediate/constrainedlayout_guide.html#legends) with `wanttoprint = True` but it gives an error `AttributeError: 'Figure' object has no attribute 'do_constrained_layout'`.

Assuming that what is meant here is `set_constrained_layout`, the problem is that the figure shown deviates from the one obtained via `wanttoprint = False`. So it seems it's not possible to show the figure after saving it?

Using the second code (the one with `fig.legend()`) everything works as expected.

In my eyes there are the following things to do:

1. Correct the first code to be runnable and produce the expected outcome for both settings of `wanttoprint` on screen and as saved image.
2. Explain the reasoning behind turning the legend in layout option first off then on etc.
3. Add a savefig line to the second example to make is clear that this is to be understood parallel to the previous example.

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Closes #12391.  

When CI passes I'll post the result here for review....

## As it'll post in the docs
 
cl

## As it looks in saved file
outname

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: tutorials/intermediate/constrainedlayout_guide.py [hunk 1]
```diff
@@ -213,23 +213,30 @@ def example_plot(ax, fontsize=12, nodec=False):
...
     leg.set_in_layout(True)
-    fig.do_constrained_layout(False)
+    # we don't want the layout to change at this point.  
+    fig.set_constrained_layout(False)
     fig.savefig('outname.png', bbox_inches='tight')
 
 #############################################
 # A better way to get around this awkwardness is to simply
-# use a legend for the figure:
+# use the legend method provided by the figure class:
```
Comment (timhoffm) [inline tutorials/intermediate/constrainedlayout_guide.py]: simply ``use `.Figure.legend`:``

**Comment by jklymak** [comment]
Can we include images in the tutorials?  I guess I could just include the png.  

WRT item 3 I was t sure it was necessary but  happy to do it if you feel it helps

**Comment by ImportanceOfBeingErnest** [comment]
I think some of the confusion comes from looking at the produced images, so if at least the code has the savefig command in it, it would point more towards where to look for the feature. 

I *think* you can use rst inside the examples, so `.. image:: some.png` might just work.

**Comment by ImportanceOfBeingErnest** [comment]
The manually added figures are a bit... HUGE. 

image


But appart this looks good. With both figures shown like this I think the whole section becomes really clear.

**Comment by jklymak** [comment]
Seems better now:

https://14431-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/tutorials/intermediate/constrainedlayout_guide.html#sphx-glr-tutorials-intermediate-constrainedlayout-guide-py

**Comment by ImportanceOfBeingErnest** [comment]
What's the reason for the difference? Shouldn't it save with the same dpi as it shows?

**Comment by jklymak** [comment]
I had to hard-set the width to 300 px in the `rst`.  We are only trying to show what the layout looks like, not what the exact sizing of the png is.

**Comment by ImportanceOfBeingErnest** [comment]
image

File: tutorials/intermediate/constrainedlayout_guide.py [hunk 2]
```diff
@@ -213,30 +212,47 @@ def example_plot(ax, fontsize=12, nodec=False):
...
 leg = fig.legend(lines, labels, loc='center left',
-                 bbox_to_anchor=(0.8, 0.5), bbox_transform=axs[1, 1].transAxes)
+                 bbox_to_anchor=(0.8, 0.5), bbox_transform=axs[1].transAxes)
+fig.savefig('CL02.png', bbox_inches='tight', dpi=100)
+
+#############################################
+# The saved file looks like:
+#
+# .. image:: /_static/constrained_layout/CL02.png
+#    :align: center
```
Comment (ImportanceOfBeingErnest) [inline tutorials/intermediate/constrainedlayout_guide.py]: The center alignment seems to be ignored for some reason by the css. I would not consider this crucial at this point. One can revisit the cause of this at a different stage.

Comment (jklymak) [reply tutorials/intermediate/constrainedlayout_guide.py]: :-(   Oh well I tried.  As I'm sure you can tell screwing around w/ rst embedded in python comments isn't my favourite pastime.

**Comment by jklymak** [comment]
Squased;  final product at https://14460-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/tutorials/intermediate/constrainedlayout_guide.html#sphx-glr-tutorials-intermediate-constrainedlayout-guide-py

File: tutorials/intermediate/constrainedlayout_guide.py (hunk 2 above)
Comment (ImportanceOfBeingErnest) [reply tutorials/intermediate/constrainedlayout_guide.py]: There is nothing wrong with the rst command itself. It's either sphinx having changed something in their newest version or us screwing up the CSS. 

I think adding something like

```
img.align-center {    
    display: block;
    margin-left: auto;
    margin-right: auto;}
```
in the css file will center the image correctly. 

It's something to keep in mind, but somehow unrelated to this PR anyways.

**Comment by jklymak** [comment]
Thanks for the help @ImportanceOfBeingErnest

**Comment by ImportanceOfBeingErnest** [comment]
It seems one drawback of this is that now every time you build the docs it'll create two png images in the tutorials folder, which will then end up in any commit (if one doesn't pay attention).

**Pull request merged**